src/mlpack/methods/regularized_svd/regularized_svd_function_impl.hpp

```diff
@@ -186,5 +218,109 @@ double StandardSGD::Optimize(
   return overallObjective;
 }
 
+
+template <>
+template <>
+inline double ParallelSGD<ExponentialBackoff>::Optimize(
+    mlpack::svd::RegularizedSVDFunction<arma::mat>& function,
+    arma::mat& iterate)
+{
+  double overallObjective = DBL_MAX;
+  double lastObjective;
+
+  // The order in which the functions will be visited.
+  arma::Col<size_t> visitationOrder = arma::linspace<arma::Col<size_t>>(0,
+      (function.NumFunctions() - 1), function.NumFunctions());
+
+  const arma::mat data = function.Dataset();
+
+  // Iterate till the objective is within tolerance or the maximum number of
+  // allowed iterations is reached. If maxIterations is 0, this will iterate
+  // till convergence.
+  for (size_t i = 1; i != maxIterations; ++i)
+  {
+    // Calculate the overall objective.
+    lastObjective = overallObjective;
+    overallObjective = 0;
+
+    #pragma omp parallel for reduction(+:overallObjective)
+    for (omp_size_t j = 0; j < (omp_size_t) function.NumFunctions(); ++j)
+    {
+      overallObjective += function.Evaluate(iterate, j);
+    }
+
+    // Output current objective function.
+    Log::Info << "Parallel SGD: iteration " << i << ", objective "
+      << overallObjective << "." << std::endl;
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "Parallel SGD: converged to " << overallObjective
+        << "; terminating with failure. Try a smaller step size?"
+        << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SGD: minimized within tolerance " << tolerance << "; "
+        << "terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    // Get the stepsize for this iteration
+    double stepSize = decayPolicy.StepSize(i);
+
+    // Shuffle for uniform sampling of functions by each thread.
+    std::random_shuffle(visitationOrder.begin(), visitationOrder.end());
```

Comment:
I agree that the absolute difference begins to make a significant affect only when n ~ 1e8. One more thing `arma::shuffle` does is to create another copy of the matrix, which is not needed.

```
λ mlpack_spike/random ∴ ./a.out
Run on (4 X 3700 MHz CPU s)
2017-07-24 01:42:06
Benchmark             Time           CPU Iterations
---------------------------------------------------
baseline     3614062113 ns 3610419547 ns          1
arma_shuffle 16072122317 ns 16071446490 ns          1
std_shuffle  8582545673 ns 8582178549 ns          1
```

I ran the benchmark I wrote for size = 1e6, and `arma::shuffle` took 3x more time as compared to `std::random_shuffle`, while consuming ~2x more memory (due to the extra copy).

Should I switch back to `arma::shuffle`?